### PR TITLE
Support for mysql schema with spatial types/indexes

### PR DESCRIPTION
**Pull request by Vasfed**
### Summary

This allows one to have spatial fields and indexes in database without switching to sql schema format.

### CODE REVIEW

**Comment by jeremy** [comment]
Cool!

**Comment by arthurnn** [comment]
When was this added to MySQL?
Should we check the MySQL version before allowing these?
Is this something other adpters have support too?

**Review by rafaelfranca — CHANGES_REQUESTED** [review]
Could you add a CHANGELOG entry?

@arthurnn I don't think we need to check the MySQL version. People needs to opt-in this feature and If people use a version that doesn't support it they will see when running the migration.

**Comment by Vasfed** [comment]
@arthurnn mysql added spatial extensions in 4.1, and as of 5.7.5 spatial indexing for InnoDB was added. Since we are supporting Mysql>=5.1.10 no additional checks are needed, except for index addition, but in this case one will get an error on migration.
Also Rails already has support for `:spatial` index type, but these were dumped incorrectly because spatial indexes cannot have prefixes, but report them (this PR fixes this).

**Comment by Vasfed** [comment]
@rafaelfranca added changelog entry + rebased onto current master to resolve conflicts

File: activerecord/lib/active_record/connection_adapters/mysql/schema_definitions.rb [hunk 1]
```diff
@@ -54,6 +54,42 @@ def unsigned_float(*args, **options)
...
+          args.each { |name| column(name, :multi_point, options) }
+        end
+
+        def multi_linestring(*args, **options)
+          args.each { |name| column(name, :multi_linestring, options) }
+        end
+
+        def multi_polygon(*args, **options)
+          args.each { |name| column(name, :multi_polygon, options) }
+        end
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/mysql/schema_definitions.rb]: Don't think it makes sense exposing those multi columns both via an option and a method, I'd go with the option across the board:

```ruby
def point(*args, multi: false, **options)
  args.each { |name| column(name, multi ? :multi_point : :point, options) }
end
```

Then the same for the other versions.

Comment (Vasfed) [inline activerecord/lib/active_record/connection_adapters/mysql/schema_definitions.rb]: @kaspth I agree. Pushed changes

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 2]
```diff
@@ -396,7 +404,7 @@ def indexes(table_name, name = nil) #:nodoc:
             end
 
             indexes.last.columns << row[:Column_name]
-            indexes.last.lengths.merge!(row[:Column_name] => row[:Sub_part].to_i) if row[:Sub_part]
+            indexes.last.lengths.merge!(row[:Column_name] => row[:Sub_part].to_i) if row[:Sub_part] && mysql_index_type != :spatial
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: @jeremy can you vet this change?

**Comment by kaspth** [comment]
Think this is the limit of what I can meaningfully contribute. Handing it over, thanks @Vasfed!

r? @jeremy

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb (hunk 2 above)
Comment (Vasfed) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: Mysql returns non-null `Sub_part` for spatial index, but does not allow to set it, thus previous behaviour resulted in a schema dump that could not be loaded back.

https://dev.mysql.com/doc/refman/5.7/en/create-index.html :

> Spatial indexes (created using SPATIAL INDEX) have these characteristics:
> ...
> - Column prefix lengths are prohibited. The full width of each column is indexed.

**Comment by jeremy** [comment]
How does this integrate with https://github.com/rgeo/rgeo or https://github.com/geokit/geokit-rails ?

I notice that other spatial adapters use distinct "multi" types rather than making it an option on the singular type, e.g. `multi_polygon :regions` instead of `polygon :regions, multi: true`. What's your feeling on the style difference?

**Comment by Vasfed** [comment]
@jeremy postgresql adapter provides similar types by adding `array: true` modifier to a spatial column. While whole thing is completely db-dependent I think less of mysql implementation details can be leaked into interface.

geokit-rails does not use native spatial extensions, only mentions mysql2spatial adapter from rgeo as a synonym for regular mysql adapter for using mysql's math functions.

Will look closer into implementation of rgeo adapter, it's master branch is incompatible with AR 5 ( rgeo/activerecord-mysql2spatial-adapter#19 ) at the moment. I doubt that all heavy-lifting of parsing binary WKB geo data should be brought into rails, so better is to provide an integration point. First thing coming to mind is monkey-patching `MysqlGeometry` class from rgeo, what is current policy on this?

**Comment by sikachu** [comment]
Looks like this PR has been opened for more than a year now. Do you think it's worth rebasing this and pushing it forward to be included in 6.0?

**Pull request closed without merging**